design/ingressroute-design.md

```diff
@@ -0,0 +1,300 @@
+This permits the _owner_ of an ingress root to both delegate the authority to publish a service on a portion of the route space inside a virtual host, and to itself further delegate authority to publish and delegate.
+
+This also means that 
+
+In practice the linkage, or delegation, from root to edge, is performed with a specific type of route action.
+You can think of it that rather than routing traffic to a service, it is routed to another ingressroute object for further processing.
+
+
+##
+
+# Detailed Design
+
+## IngressRoute CRD
+
+This is an example of a fully populated root ingressroute.
+
+```yaml
+apiVersion: contour.heptio.com/v1alpha1
+kind: IngressRoute
+metadata:
+  name: google
+  namespace: prod
+spec:
+  # virtualhost appears at most once. If it is present, the object is considered
+  # to be a "root".
+  virtualhost:
+    # the fully qualified domain name of the root of the ingress tree
+    # all leaves of the DAG rooted at this object relate to the fqdn (and its aliases)
+    fqdn: www.google.com
+    # a set of aliases for the domain, these may be alternative fqdns which are considered
+    # aliases of the primary fqdn
+    aliases:
+      - www.google.com.au
+      - google.com
+    # if present describes tls properties. The CNI names that will be matched on
+    # are described in fqdn and aliases, the tls.secretName secret must contain a
+    # matching certificate
+    tls:
+      # required, the name of a secret in the current namespace
+      secretName: google-tls
+      # other properties like cipher suites may be added later
+  # routes contains the set of routes for this virtual host.
+  # routes must _always_ be present and non empty.
+  # routes can be present in any order, and will be matched from most to least
+  # specific.
+  routes:
+  # each route entry starts with a prefix match
+  # and one of service or delegate
+  - match: /static
+    # service defines the properties of the service in the current namespace
+    # that will handle traffic matching the route
+    service:
+    - name: google-static
+      port: 9000
+  - match: /finance
+    # delegate delegates the matching route to another IngressRoute object.
+    # This delegates the responsibility for /finance to the IngressRoute matching the delegate parameters
+    delegate:
+      # the name of the IngressRoute object to delegate to
+      name: google-finance
+      # the namespace of the IngressRoute object, if blank the current namespace is assumed
+      namespace: finance
+  - match: /ads
+    # more than one service name/port pair may be present allowing the services for a route to
+    # be handled by more than one service.
+    service:
+    - name: ads-red
+      port: 9090
+    - name: ads-blue
+      port: 9090
+```
+
+This is an example of the `google-finance` object which has been delegated responsibilty for paths starting with `/finance`.
```

Comment:
I imagine there are two modes. One is your usual kubernetes free for all were all configuration pushed to the cluster is "trusted"; either the cluster is small enough that there are few operators, or the changes to the cluster go through some CI process to "catch" rogue changes.

The other mode could be some kind of "enforcing" mode where contour will only look for roots in a set of namespaces. So new roots can be created, but will be ignored outside approved. Ingress adminstrators would create root objects in their namespace and use "complete?" delegation to delegate "/" to another object in another namespace.

Reply:
There are two dimensions of permission / RBAC / whatever. The first is using a pair of objects to imply tiers of permission, aka the unix model with root and non root; we have root ingress routes and non root ingress routes. The other is using the moral equivalent of file permissions; my user can write to /etc, yours can't, so while it's all a file, i can edit files you cannot.

The model I have in my head, which i hope mirrors joe's, is the latter; a flag or optional configuration that means contour will ignore roots except in a set of nominated namespaces. @jbeda I have not called this out in the document because this enforcement mode is out of scope for this design.

Reply:
@davecheney @jbeda I believe that enforcement mode should be in scope for this design. That said, I agree with the model of having a flag to ignore roots outside a set of nominated namespaces.